### ISSUE TITLE
Colour palette improvements

### ISSUE THREAD

**Issue by kmcnaught** [microbit-foundation/ml-trainer#510]
(moving discussion here from email). 

We now have 2 colourblind palettes (yay). 

For next steps, consider:

1. a line width option
2. an option to use dots/dashes for more visual differentiation. This could be something like:
  - a checkbox for "line contrast" or "line patterns" (am struggling with wording which might suggest this isn't right...)
  - a "Line Style" dropdown for "solid lines" (default) vs "dots and dashes" or similar
3. For the colour blind palettes, consider how to work better with Windows High Contrast Mode - see #509

**Comment by microbit-robert** [comment]
@kmcnaught We had a go at modifying the library that manages the live data streaming to enable styling the lines using dots /dashes. We've hit an interesting problem which is shown in the attached video. When the lines reach the left hand edge of the canvas, old data is discarded and the lines are redrawn. The dotted and dashed lines are redrawn from the left edge (they always start with a dot or dash rather than preserving the previous state which could have been a space between a dash, half a dash, etc.) and, as new data is added, it gives the impression that the styled lines are moving which is quite distracting to say the least.

Do you know if this is a common problem with existing workarounds? We can reduce how often we discard old data so that these styles only appear to move every one or two seconds, but this is still not particularly satisfying. 

https://github.com/user-attachments/assets/d0e75e50-b898-4640-ad7e-d8f9999ff72e

**Comment by kmcnaught** [comment]
Ah, I see the problem! I haven't seen it before (but is presumably going to be an issue any time anyone uses dashed lines for live data..). Is there a concept of `stroke-dashoffset` (like in CSS/SVG) in the library you are using? That could be updated in sync with the rate you drop data.

**Comment by microbit-robert** [comment]
Yes, I think an issue for any live data using dashed lines. We've seen the same thing in Plotly and a streaming data plugin for ChartJS as well.

There's a `lineDashOffset` property, but it seems very difficult to calculate such an offset as there's no convient map how many pixels a data point represents, though I might have another go at this.

**Comment by microbit-robert** [comment]
@kmcnaught We think we've now cracked this. Can you see any potential issues if we follow through with an implementation that leads to these styles of lines for both the recording and live graphs in the screenshot below?

![image](https://github.com/user-attachments/assets/883d9672-237f-48bc-ab31-2d99f8838d60)

**Comment by kmcnaught** [comment]
Nice one. That looks good to me.

**Comment by microbit-robert** [comment]
See https://github.com/microbit-foundation/ml-trainer/pull/548
